### PR TITLE
Fix StringReader.read

### PR DESCRIPTION
**Pull request by hastri**
#6142

### CODE REVIEW

**Comment by hastri** [comment]
Adds isEof to /std/io/ioutils.ts

**Comment by ry** [comment]
Can you add a test for the bug that is being fixed?

**Comment by hastri** [comment]
That's the first commit, without the fix it fails

**Comment by ry** [comment]
it tests isEof - I'm looking for a test demonstrating a bug in the existing API

**Comment by hastri** [comment]
It kind of does, I just extracted it to reduce code duplication. Should I add a test with the code inlined?

**Comment by hastri** [comment]
The problem is that StringReader.read always returns EOF if the passed array is empty

**Comment by ry** [comment]
> The problem is that StringReader.read always returns EOF if the passed array is empty

Sounds like a good test

File: std/io/ioutil_test.ts [hunk 1]
```diff
@@ -98,3 +100,22 @@ Deno.test("copyNWriteAllData", async function (): Promise<void> {
...
   assertEquals(n, size);
 });
+
+Deno.test("testStringReaderEof", async function (): Promise<void> {
+  const r = new StringReader("abc");
+  assertEquals(await r.read(new Uint8Array()), 0);
+  assertEquals(await r.read(new Uint8Array(4)), 3);
+  assertEquals(await r.read(new Uint8Array(1)), null);
+  assertEquals(await r.read(new Uint8Array(0)), null);
+});
```
Comment (ry) [inline std/io/ioutil_test.ts:110]: 👍

File: std/io/ioutil.ts [hunk 2]
```diff
@@ -82,3 +83,9 @@ export function sliceLongToBytes(d: number, dest = new Array(8)): number[] {
   }
   return dest;
 }
+
+/** Checks if `r` reached EOF */
+export async function isEof(r: Deno.Reader): Promise<boolean> {
+  const read = await r.read(empty);
+  return read == null;
+}
```
Comment (ry) [inline std/io/ioutil.ts:91]: Seems like it might not be necessary to add this after all?

File: std/io/ioutil_test.ts [hunk 3]
```diff
@@ -98,3 +100,22 @@ Deno.test("copyNWriteAllData", async function (): Promise<void> {
 
   assertEquals(n, size);
 });
+
+Deno.test("testStringReaderEof", async function (): Promise<void> {
+  const r = new StringReader("abc");
```
Comment (ry) [inline std/io/ioutil_test.ts:105]: use `stringsReader("abc")`

File: std/io/ioutil_test.ts [hunk 4]
```diff
@@ -8,8 +8,10 @@ import {
   readLong,
   readShort,
   sliceLongToBytes,
+  isEof,
 } from "./ioutil.ts";
 import { BufReader } from "./bufio.ts";
+import { StringReader } from "./readers.ts";
```
Comment (ry) [inline std/io/ioutil_test.ts:14]: remove

File: std/io/ioutil_test.ts [hunk 5]
```diff
@@ -98,3 +100,22 @@ Deno.test("copyNWriteAllData", async function (): Promise<void> {
...
+Deno.test("testIsEof", async function (): Promise<void> {
+  const r = new StringReader("abc");
+  assertEquals(await isEof(r), false);
+  await r.read(new Uint8Array(2));
+  assertEquals(await isEof(r), false);
+  await r.read(new Uint8Array(2));
+  assertEquals(await isEof(r), true);
+  await r.read(new Uint8Array(2));
+  assertEquals(await isEof(r), true);
+});
```
Comment (ry) [inline std/io/ioutil_test.ts:121]: if possible remove isEof and this test

File: std/io/ioutil_test.ts (hunk 3 above)
Comment (hastri) [reply std/io/ioutil_test.ts:105]: But then it won't test StringReader, which has the bug

Comment (ry) [reply std/io/ioutil_test.ts:105]: Ah I see. Somehow this StringReader class was added erroneously. The correct implementation is stringsReader.

Comment (hastri) [reply std/io/ioutil_test.ts:105]: maybe test with both StringReader and stringsReader(Buffer)?

Comment (hastri) [reply std/io/ioutil_test.ts:105]: So the StringReader should probably just be removed

Comment (hastri) [reply std/io/ioutil_test.ts:105]: and replaced with stringsReader once every occurrence

**Comment by ry** [comment]
StringReader should never have been added - it's a bad implementation. The correct implementation is stringsReader. closing in favor of https://github.com/denoland/deno/pull/6148

**Pull request closed without merging**